Inconsistent goal output

On Pants main @ 6037e699 I encountered this failure with no information about why Flake8 failed:
```console
$ ./pants --changed-since=upstream/main fmt lint check test
10:47:55.31 [INFO] Initializing scheduler...
10:47:55.41 [INFO] Scheduler initialized.
10:47:58.52 [INFO] Completed: Building autoflake.pex from autoflake_default_lockfile.txt
10:47:58.64 [WARN] Completed: Format with Autoflake - autoflake made changes.
10:48:03.67 [INFO] Completed: Building black.pex from black_default_lockfile.txt
10:48:04.21 [WARN] Completed: Format with Black - Black made changes.
reformatted src/python/pants/engine/process_test.py
All done! ✨ 🍰 ✨
1 file reformatted, 1 file left unchanged.


10:48:07.37 [INFO] Completed: Building docformatter.pex from docformatter_default_lockfile.txt
10:48:07.48 [INFO] Completed: Format with docformatter - Docformatter made no changes.
10:48:08.47 [INFO] Completed: Building isort.pex from isort_default_lockfile.txt
10:48:08.56 [INFO] Completed: Format with isort - isort made no changes.

+ Black made changes.
✓ Docformatter made no changes.
+ autoflake made changes.
✓ isort made no changes.
10:48:08.65 [INFO] Completed: Lint with isort - isort succeeded.
10:48:08.68 [INFO] Completed: Lint with autoflake - autoflake succeeded.
10:48:08.73 [INFO] Completed: Lint with docformatter - Docformatter succeeded.
10:48:09.02 [INFO] Completed: Lint with Black - Black succeeded.
All done! ✨ 🍰 ✨
2 files would be left unchanged.


10:48:11.36 [INFO] Completed: Building flake8.pex from 3rdparty/python/lockfiles/flake8.txt

✓ Black succeeded.
✓ Docformatter succeeded.
𐄂 Flake8 failed.
✓ autoflake succeeded.
✓ isort succeeded.
```

I re-ran and found:
```console
$ ./pants --changed-since=upstream/main fmt lint check test
10:49:05.17 [INFO] Completed: Format with Autoflake - autoflake made no changes.
10:49:05.56 [INFO] Completed: Format with Black - Black made no changes.
All done! ✨ 🍰 ✨
2 files left unchanged.


10:49:05.56 [INFO] Completed: Format with docformatter - Docformatter made no changes.
10:49:05.56 [INFO] Completed: Format with isort - isort made no changes.

✓ Black made no changes.
✓ Docformatter made no changes.
✓ autoflake made no changes.
✓ isort made no changes.
10:49:05.57 [INFO] Completed: Lint with autoflake - autoflake succeeded.
10:49:05.57 [INFO] Completed: Lint with docformatter - Docformatter succeeded.
10:49:05.57 [INFO] Completed: Lint with Black - Black succeeded.
All done! ✨ 🍰 ✨
2 files would be left unchanged.


10:49:05.57 [INFO] Completed: Lint with isort - isort succeeded.
10:49:05.77 [ERROR] Completed: Lint with Flake8 - Flake8 failed (exit code 1).
src/python/pants/engine/process.py:537:1: W293 blank line contains whitespace



✓ Black succeeded.
✓ Docformatter succeeded.
𐄂 Flake8 failed.
✓ autoflake succeeded.
✓ isort succeeded.
```

So the issue was `src/python/pants/engine/process.py:537:1: W293 blank line contains whitespace` but that was not reported on the 1st run. In the 1st run it's just reported that the flake8.pex was built `10:48:11.36 [INFO] Completed: Building flake8.pex from 3rdparty/python/lockfiles/flake8.txt`, but not what the `Completed: ...` of that pex's execution was.

The pantsd.log shows the following for this session:
```
10:47:54.88 [INFO] handling request: `--changed-since=upstream/main fmt lint check test`
10:48:08.57 [INFO] notify invalidation: cleared 2 and dirtied 109 nodes for: {"src/python/pants/engine/process.py", "src/python/pants/engine"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine/process.py", "src/python/pants/engine"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process.py"}
10:48:08.57 [INFO] notify invalidation: cleared 1 and dirtied 15 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process_test.py"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine/process_test.py", "src/python/pants/engine"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process_test.py"}
10:48:08.57 [INFO] notify invalidation: cleared 2 and dirtied 111 nodes for: {"src/python/pants/engine/process.py", "src/python/pants/engine"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process_test.py"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine/process.py", "src/python/pants/engine"}
10:48:08.57 [INFO] Filesystem changed during run: retrying `PySnapshot` in 500ms...
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process.py"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process_test.py"}
10:48:08.57 [INFO] notify invalidation: cleared 0 and dirtied 0 nodes for: {"src/python/pants/engine", "src/python/pants/engine/process_test.py"}
10:48:11.56 [INFO] request completed: `--changed-since=upstream/main fmt lint check test`
10:49:04.98 [INFO] handling request: `--changed-since=upstream/main fmt lint check test`
10:49:05.77 [INFO] request completed: `--changed-since=upstream/main fmt lint check test`
10:49:15.41 [INFO] Extending leases
10:49:15.43 [INFO] Done extending leases
...
```

And one last time with test - this is very easy to repro apparently:
```console
$ ./pants --changed-since=upstream/main fmt lint check test
11:10:10.73 [INFO] Completed: Format with Autoflake - autoflake made no changes.
11:10:11.19 [WARN] Completed: Format with Black - Black made changes.
reformatted src/python/pants/engine/process_test.py
All done! ✨ 🍰 ✨
1 file reformatted, 1 file left unchanged.


11:10:11.30 [INFO] Completed: Format with docformatter - Docformatter made no changes.
11:10:11.36 [INFO] Completed: Format with isort - isort made no changes.

+ Black made changes.
✓ Docformatter made no changes.
✓ autoflake made no changes.
✓ isort made no changes.
11:10:11.37 [INFO] Filesystem changed during run: retrying `Fmt` in 500ms...
11:10:11.95 [INFO] Completed: Lint with isort - isort succeeded.
11:10:11.98 [INFO] Completed: Lint with autoflake - autoflake succeeded.
11:10:12.00 [INFO] Completed: Lint with docformatter - Docformatter succeeded.
11:10:12.08 [INFO] Completed: Lint with Flake8 - Flake8 succeeded.
11:10:12.28 [INFO] Completed: Lint with Black - Black succeeded.
All done! ✨ 🍰 ✨
2 files would be left unchanged.



✓ Black succeeded.
✓ Docformatter succeeded.
✓ Flake8 succeeded.
✓ autoflake succeeded.
✓ isort succeeded.
11:10:16.98 [INFO] Completed: Typecheck using MyPy - MyPy succeeded.
Success: no issues found in 2 source files



✓ MyPy succeeded.
11:10:22.63 [INFO] Completed: Building pytest.pex from 3rdparty/python/lockfiles/pytest.txt
11:10:22.87 [INFO] Completed: Building local_dists.pex
11:10:23.84 [INFO] Completed: Building pytest_runner.pex

𐄂 src/python/pants/engine/process_test.py:tests failed.
```

Fine on 2nd run though:
```console
$ ./pants --changed-since=upstream/main fmt lint check test
11:10:40.58 [INFO] Completed: Format with Autoflake - autoflake made no changes.
11:10:40.97 [INFO] Completed: Format with Black - Black made no changes.
All done! ✨ 🍰 ✨
2 files left unchanged.


11:10:40.97 [INFO] Completed: Format with docformatter - Docformatter made no changes.
11:10:40.97 [INFO] Completed: Format with isort - isort made no changes.

✓ Black made no changes.
✓ Docformatter made no changes.
✓ autoflake made no changes.
✓ isort made no changes.
11:10:40.98 [INFO] Completed: Lint with Black - Black succeeded.
All done! ✨ 🍰 ✨
2 files would be left unchanged.


11:10:40.98 [INFO] Completed: Lint with autoflake - autoflake succeeded.
11:10:40.98 [INFO] Completed: Lint with isort - isort succeeded.
11:10:40.98 [INFO] Completed: Lint with docformatter - Docformatter succeeded.
11:10:40.98 [INFO] Completed: Lint with Flake8 - Flake8 succeeded.

✓ Black succeeded.
✓ Docformatter succeeded.
✓ Flake8 succeeded.
✓ autoflake succeeded.
✓ isort succeeded.
11:10:41.04 [INFO] Completed: Typecheck using MyPy - MyPy succeeded.
Success: no issues found in 2 source files



✓ MyPy succeeded.
11:10:42.03 [ERROR] Completed: Run Pytest - src/python/pants/engine/process_test.py:tests failed (exit code 1).
============================= test session starts ==============================
collected 16 items

src/python/pants/engine/process_test.py ...............F                 [100%]

=================================== FAILURES ===================================
__________________________ test_find_binary_file_path __________________________

rule_runner = RuleRunner(build_root=/tmp/_BUILD_ROOTl3ikv17s)
tmpdir = local('/tmp/pytest-of-jsirois/pytest-1/test_find_binary_file_path0')

    def test_find_binary_file_path(rule_runner: RuleRunner, tmpdir: py.path.local) -> None:
        binary_name = "mybin"
        binary_path_abs = tmpdir / "bin" / binary_name
        binary_path_abs.write("", ensure=True)
        binary_path_abs.chmod(0o755)
    
        binary_paths = rule_runner.request(
            BinaryPaths,
            [BinaryPathRequest(binary_name=binary_name, search_path=[binary_path_abs.strpath])],
        )
>       assert binary_paths.first_path is not None
E       AssertionError: assert None is not None
E        +  where None = BinaryPaths(binary_name='mybin', paths=()).first_path

src/python/pants/engine/process_test.py:315: AssertionError
- generated xml file: /tmp/process-executionHDL5Ut/src.python.pants.engine.process_test.py.tests.xml -
=========================== short test summary info ============================
FAILED src/python/pants/engine/process_test.py::test_find_binary_file_path - ...
========================= 1 failed, 15 passed in 0.75s =========================



𐄂 src/python/pants/engine/process_test.py:tests failed.
```

--- a/src/rust/engine/ui/Cargo.toml
+++ b/src/rust/engine/ui/Cargo.toml
@@ -13,6 +13,7 @@ futures = "0.3"
 indexmap = "1.4"
 indicatif = "0.16.2"
 logging = { path = "../logging" }
+parking_lot = "0.11"
 stdio = { path = "../stdio" }
 task_executor = { path = "../task_executor" }
 uuid = { version = "0.7", features = ["v4"] }

--- a/src/rust/engine/ui/src/console_ui.rs
+++ b/src/rust/engine/ui/src/console_ui.rs
@@ -88,6 +88,9 @@ impl ConsoleUI {
     // We take exclusive access to stdio by registering a callback that is used to render stderr
     // while we're holding access. See the method doc.
     let stderr_dest_bar: Arc<Mutex<Option<WeakProgressBar>>> = Arc::new(Mutex::new(None));
+    // We acquire the lock before taking exclusive access, and don't release it until after
+    // initialization. That way, the exclusive callback can always assume that the destination
+    // is initialized (i.e., can `unwrap` it).
     let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
     let (term_read, _, term_stderr_write) = {
       let stderr_dest_bar = stderr_dest_bar.clone();

--- a/src/rust/engine/ui/src/console_ui.rs
+++ b/src/rust/engine/ui/src/console_ui.rs
@@ -27,12 +27,13 @@
 
 use std::future::Future;
 use std::pin::Pin;
-use std::sync::mpsc;
+use std::sync::Arc;
 use std::time::Duration;
 
 use futures::future::{self, FutureExt, TryFutureExt};
 use indexmap::IndexMap;
-use indicatif::{MultiProgress, ProgressBar, ProgressDrawTarget, ProgressStyle};
+use indicatif::{MultiProgress, ProgressBar, ProgressDrawTarget, ProgressStyle, WeakProgressBar};
+use parking_lot::Mutex;
 
 use task_executor::Executor;
 use workunit_store::{format_workunit_duration, SpanId, WorkunitStore};
@@ -76,18 +77,33 @@ impl ConsoleUI {
   ///
   /// Setup progress bars, and return them along with a running task that will drive them.
   ///
+  /// NB: This method must be very careful to avoid logging. Between the point where we have taken
+  /// exclusive access to the Console and when the UI has actually been initialized, attempts to
+  /// log from this method would deadlock (by causing the method to wait for _itself_ to finish).
+  ///
   fn setup_bars(
     &self,
     executor: Executor,
   ) -> Result<(Vec<ProgressBar>, MultiProgressTask), String> {
-    // Stderr is propagated across a channel to remove lock interleavings between stdio and the UI.
-    let (stderr_sender, stderr_receiver) = mpsc::sync_channel(0);
-    let (term_read, _, term_stderr_write) =
+    // We take exclusive access to stdio by registering a callback that is used to render stderr
+    // while we're holding access. See the method doc.
+    let stderr_dest_bar: Arc<Mutex<Option<WeakProgressBar>>> = Arc::new(Mutex::new(None));
+    let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
+    let (term_read, _, term_stderr_write) = {
+      let stderr_dest_bar = stderr_dest_bar.clone();
       stdio::get_destination().exclusive_start(Box::new(move |msg: &str| {
-        // If we fail to send, it's because the UI has shut down: we fail the callback to
-        // have the logging module directly log to stderr at that point.
-        stderr_sender.send(msg.to_owned()).map_err(|_| ())
-      }))?;
+        // Acquire a handle to the destination bar in the UI. If we fail to upgrade, it's because
+        // the UI has shut down: we fail the callback to have the logging module directly log to
+        // stderr at that point.
+        let dest_bar = {
+          let stderr_dest_bar = stderr_dest_bar.lock();
+          // We can safely unwrap here because the Mutex is held until the bar is initialized.
+          stderr_dest_bar.as_ref().unwrap().upgrade().ok_or(())?
+        };
+        dest_bar.println(msg);
+        Ok(())
+      }))?
+    };
 
     let stderr_use_color = term_stderr_write.use_color;
     let term = console::Term::read_write_pair_with_style(
@@ -106,21 +122,12 @@ impl ConsoleUI {
         multi_progress.add(ProgressBar::new(50).with_style(style))
       })
       .collect::<Vec<_>>();
-    let first_bar = bars[0].downgrade();
+    *stderr_dest_bar_guard = Some(bars[0].downgrade());
 
     // Spawn a task to drive the multi progress.
     let multi_progress_task = executor
       .spawn_blocking(move || multi_progress.join())
       .boxed();
-    // And another to propagate stderr, which will exit automatically when the channel closes.
-    let _stderr_task = executor.spawn_blocking(move || {
-      while let Ok(stderr) = stderr_receiver.recv() {
-        match first_bar.upgrade() {
-          Some(first_bar) => first_bar.println(stderr),
-          None => break,
-        }
-      }
-    });
 
     Ok((bars, multi_progress_task))
   }

--- a/src/rust/engine/ui/src/console_ui.rs
+++ b/src/rust/engine/ui/src/console_ui.rs
@@ -42,7 +42,6 @@ pub struct ConsoleUI {
   local_parallelism: usize,
   // While the UI is running, there will be an Instance present.
   instance: Option<Instance>,
-  teardown_mpsc: (mpsc::Sender<()>, mpsc::Receiver<()>),
 }
 
 impl ConsoleUI {
@@ -51,7 +50,6 @@ impl ConsoleUI {
       workunit_store,
       local_parallelism,
       instance: None,
-      teardown_mpsc: mpsc::channel(),
     }
   }
 
@@ -207,16 +205,11 @@ impl ConsoleUI {
   ///
   pub fn teardown(&mut self) -> impl Future<Output = Result<(), String>> {
     if let Some(instance) = self.instance.take() {
-      let sender = self.teardown_mpsc.0.clone();
       // When the MultiProgress completes, the Term(Destination) is dropped, which will restore
       // direct access to the Console.
       instance
         .multi_progress_task
         .map_err(|e| format!("Failed to render UI: {}", e))
-        .and_then(move |()| {
-          sender.send(()).unwrap();
-          future::ok(())
-        })
         .boxed()
     } else {
       future::ok(()).boxed()

